lib/roken/getauxval.c

```diff
@@ -0,0 +1,308 @@
+    if (ret != 0 || errno == ENOENT || getauxval_sets_errno == 1) {
+        if (ret != 0)
+            errno = save_errno;
+        else if (getauxval_sets_errno > 0 && errno == 0)
+            errno = save_errno;
+        return ret;
+    }
+
+    if (getauxval_sets_errno == 0) {
+        errno = save_errno;
+        if ((a = rk_getprocauxv(type)) == NULL) {
+            errno = ENOENT;
+            return 0;
+        }
+        return a->a_un.a_val;
+    }
+
+    /*
+     * We've called getauxval() and it returned 0, but we don't know if
+     * getauxval() sets errno = ENOENT when entries are not found.
+     *
+     * Attempt to detect whether getauxval() sets errno = ENOENT by
+     * calling it with what should be a bogus type.
+     */
+
+    errno = 0;
+    ret2 = getauxval(~type);
+    if (ret2 == 0 && errno == ENOENT) {
+        getauxval_sets_errno = 1;
+        errno = save_errno;
+        return ret;
+    }
+
+    getauxval_sets_errno = 0;
+    errno = save_errno;
+#endif
+#else
+    const auxv_t *a;
+
+    if ((a = rk_getprocauxv(type)) == NULL) {
+        errno = ENOENT;
+        return 0;
+    }
+    return a->a_un.a_val;
+#endif
+}
+
+/**
+ * Like the nearly-standard getauxval().  If the auxval is not found
+ * returns NULL and always sets errno to ENOENT.  Otherwise if auxval is
+ * found it leaves errno as it was, even if the value is zero.
+ *
+ * @return The value of the ELF auxiliary value for the given type, or
+ * NULL and sets errno to ENOENT.
+ */
+ROKEN_LIB_FUNCTION void * ROKEN_LIB_CALL
+rk_getauxptr(unsigned long type)
+{
+#ifdef HAVE_GETAUXVAL
+    uintptr_t p = rk_getauxval(type);
+    return (void *)p;
+#else
+    const auxv_t *a;
+
+    if ((a = rk_getprocauxv(type)) == NULL) {
+        errno = ENOENT;
+        return 0;
+    }
+    return a->a_un.a_ptr;
+#endif
+}
+
+/**
+ * Like the nearly-standard getauxval().  If the auxval is not found
+ * returns NULL and always sets errno to ENOENT.  Otherwise if auxval is
+ * found it leaves errno as it was, even if the value is zero.
+ *
+ * @return The value of the ELF auxiliary value for the given type, or
+ * NULL and sets errno to ENOENT.
+ */
+ROKEN_LIB_FUNCTION void (*
+rk_getauxfnc(unsigned long type))(void) ROKEN_LIB_CALL
+{
+#ifdef HAVE_GETAUXVAL
+    uintptr_t p = rk_getauxval(type);
+    return (void (*)(void))p;
+#else
+    const auxv_t *a;
+
+    if ((a = rk_getprocauxv(type)) == NULL) {
+        errno = ENOENT;
+        return 0;
+    }
+    return a->a_un.a_fnc;
+#endif
+}
+
+/**
+ * *Internal* function for testing by injecting an ELF auxiliary vector
+ * entry.
+ *
+ * @return zero on success or ENOSPC if there are too many ELF auxiliary
+ * entries.
+ */
+ROKEN_LIB_FUNCTION int ROKEN_LIB_CALL
+rk_injectprocauxv(auxv_t *e)
+{
+    size_t i;
+    int ret;
+
+    /*
+     * This function is racy, but as an internal function never meant to
+     * be called in a threaded program, we don't care.
+     */
+
+    if (e->a_type == 0) {
```

Comment:
This is about truncating the auxv to be empty.  But we don't need nor use this.  And anyways, this should follow the call to `readprocauxv()`.